### ISSUE TITLE
Encoder layers cannot learn meaningful things

### ISSUE THREAD

**Issue by kingofoz** [siavashk/imagenet-autoencoder#2]
![interesting](https://cloud.githubusercontent.com/assets/9476604/14163088/734a7808-f724-11e5-80e5-d80b235cdbcf.png)

Please see the attached screenshot and focus on the things in the red rectangle. The left one are the weights of decoder filters in the last layer. The right one are the weights of encoder filters in the first layer. They get updated every epoch. I added below display code to https://github.com/siavashk/imagenet-autoencoder/blob/master/scripts/train.lua#L44:

  eweight = ae.net.modules[1].weight
  dweight = ae.net.modules[19].weight
  eweight_view = eweight:float():view(12, 3, 5, 5)
  dweight_view = dweight:float():view(12, 3, 5, 5)
  de = image.toDisplayTensor{input=eweight_view,
                             padding=2,
                             nrow=math.floor(math.sqrt(12)),
                             symmetric=true}
  dd = image.toDisplayTensor{input=dweight_view,
                             padding=2,
                             nrow=math.floor(math.sqrt(12)),
                             symmetric=true}
  _win1_ = image.display{image=dd, win=_win1_, legend='Decoder filters', zoom=2}
  _win2_ = image.display{image=de, win=_win2_, legend='Encoder filters', zoom=2}

I change kernel size to 5 so we can get a clearer looking. You can find the decoder layer learned something meaningful, but the encoder layer learned just noise. If I stop the training at this time and test the net. I am pretty sure the learned image would be very similar with the original one according to my previous experiments. Can you give me a little comment on this? Is the encoder part normal?

**Comment by siavashk** [comment]
As long as the training and validation errors are decreasing, I would say that learning is happening. It could be that the input to the decoded layer has the bulk of the information, which is more robust to noise. As a result, filters are more stable and exhibit more robust patterns.

**Comment by kingofoz** [comment]
I see. If I want to extract features from the bottle neck of the net, should I extract from the last encoder layer or the first decoder layer in this situation?

**Comment by kingofoz** [comment]
@siavashk Can you please give some comments about my question above? Thanks!

**Comment by kingofoz** [comment]
Let me paste the four lines:
self.net:add(nn.Reshape(24 \* 16 \* 16))
self.net:add(nn.Linear(24 \* 16 \* 16, 1568))
self.net:add(nn.Linear(1568, 24 \* 16 \* 16))
self.net:add(nn.Reshape(24, 16, 16))

Which line should I use to extract features?

**Comment by siavashk** [comment]
@kingofoz I am on vacation in Mexico until Friday, and I don't have regular access to internet. 
To answer your question, the encoded features qould be the output of this layer:
'self.net:add(nn.Linear(24 \* 16 \* 16, 1568))'

**Comment by kingofoz** [comment]
Thanks @siavashk Let me try to do some experiments.
